src/tape.jl

```diff
@@ -0,0 +1,171 @@
+using MacroTools
+import MacroTools: splitdef, combinedef
+
+mutable struct TapeNode
+    op::Function
+    args::Vector
+    results::Vector
+    kwargs::Dict
+end
+
+TapeNode(op, args, results; kwargs...) =  TapeNode(op, args, results, kwargs)
+
+mutable struct Tape
+    nodes::Dict{TensorHandle, TapeNode}
+    attrs::Dict
+end
+
+Tape(;kwargs...) = Tape(Dict{TensorHandle, TapeNode}(), Dict(kwargs...))
+
+function set_tape(new_tape=nothing)
+    if new_tape === nothing
+        new_tape = Tape()
+    end
+    context = Context()
+    context.attrs["tape"] = new_tape
+    push!(global_context, context)
+    return new_tape
+end
+
+function get_tape()
```

Comment:
can and thus  should  be a 1 line function

Reply:
not applicable anymore since it's no longer a single expression. although if there's a way to write it more compactly, then we should do it.